### PR TITLE
Refactor media to simplify line creation

### LINKED ISSUE #651: Creating a line with characteristic impedance of the media is confusing
At very low frequencies characteristic impedance of transmission line is complex. skrf doesn't seem to handle it correctly because mixing of port reference impedance `z0` and characteristic impedance of line `Z0` in calculating the S-parameters and use of powerwave S-parameter definition.

Below is example of coaxial line with 1 cm length and very low conductivity resulting in high loss. DC resistance is about 50 ohms and the frequency is 1 MHz so the line is electrically very short and could be modeled as series resistor. Requires #649 for resistance to be calculated correctly.

`media_z0` is `z0` given to the media. Currently this means setting the port impedance `z0` to this value and characteristic impedance `Z0` is left unchanged. None sets port impedance to characteristic impedance.

`line_z0` is the characteristic and port impedance to use when creating a line.

If `embed` is True then the port impedances are renormalized to `z0`.

Given the above definitions I would assume that setting `media_z0=50, line_z0=None, embed=True` should make a line with the complex characteristic impedance and renormalize the ports to 50 ohms matching the series resistor calculation. However this is not the case.

```python
import numpy as np
import skrf
from skrf.media import Coaxial, Freespace

freq = skrf.Frequency(1, 1, 1, unit='MHz')
line_l = 1e-2 # meters

for media_z0 in [50, None]:
    for line_z0 in [None, 50]:
        for line_embed in [False, True]:
            print(f'media_z0: {media_z0}, line_z0: {line_z0}, Embed: {line_embed}')
            media = Coaxial(freq, z0=media_z0, Dint=0.5e-3, Dout=1.0e-3, sigma=1e3)
            print(' z0: {:.2f}, Z0: {:.2f}, gamma: {:.2f}'.format(media.z0[0], media.Z0[0], media.gamma[0]))
            line = media.line(line_l, 'm', embed=line_embed, z0=line_z0)
            print(' line S11', line.s[0,0,0])
            print(' line S21', line.s[0,1,0])
            print(' line z0', line.z0[0,0])
            if line.z0[0,0] != 50:
                line.renormalize(50, s_def='traveling')
                print(' line renorm S11', line.s[0,0,0])
                print(' line renorm S21', line.s[0,1,0])
            print()

print('S-parameters calculated as series resistor')
r_dc = line_l*(1/media.sigma)/(np.pi*media.a**2)
print('dc R', r_dc)

if 0:
    # Manually assemble series resistor S-parameters
    y = np.zeros((1,2,2), dtype=complex)
    y[:,0,0] = 1/r_dc
    y[:,1,1] = 1/r_dc
    y[:,0,1] = -1/r_dc
    y[:,1,0] = -1/r_dc

    # This crashes
    net_r = skrf.Network(y=y, frequency=freq, z0=50)
else:
    media = Freespace(frequency=freq, z0=50)
    net_r = media.resistor(r_dc)

print('Resistor S11', net_r.s[0,0,0])
print('Resistor S21', net_r.s[0,1,0])
```

```
media_z0: 50, line_z0: None, Embed: False
 z0: 50.00, Z0: 2257.60-2257.22j, gamma: 1.14+1.14j
 line S11 0j
 line S21 (0.988617463159112-0.011255830151580522j)
 line z0 (50+0j)

media_z0: 50, line_z0: None, Embed: True
 z0: 50.00, Z0: 2257.60-2257.22j, gamma: 1.14+1.14j
 line S11 0j
 line S21 (0.988617463159112-0.011255830151580522j)
 line z0 (50+0j)

media_z0: 50, line_z0: 50, Embed: False
 z0: 50.00, Z0: 2257.60-2257.22j, gamma: 1.14+1.14j
 line S11 0j
 line S21 (0.988617463159112-0.011255830151580522j)
 line z0 (50+0j)

media_z0: 50, line_z0: 50, Embed: True
 z0: 50.00, Z0: 2257.60-2257.22j, gamma: 1.14+1.14j
 line S11 0j
 line S21 (0.988617463159112-0.011255830151580522j)
 line z0 (50+0j)

media_z0: None, line_z0: None, Embed: False
 z0: 2257.60-2257.22j, Z0: 2257.60-2257.22j, gamma: 1.14+1.14j
 line S11 0j
 line S21 (0.988617463159112-0.011255830151580522j)
 line z0 (2257.6025298862933-2257.2199592444235j)
 line renorm S11 (0.3394831573782906-9.291482605883979e-05j)
 line renorm S21 (0.6605168053866275-0.00015923170778380224j)

media_z0: None, line_z0: None, Embed: True
 z0: 2257.60-2257.22j, Z0: 2257.60-2257.22j, gamma: 1.14+1.14j
 line S11 0j
 line S21 (0.988617463159112-0.011255830151580522j)
 line z0 (2257.6025298862933-2257.2199592444235j)
 line renorm S11 (0.3394831573782906-9.291482605883979e-05j)
 line renorm S21 (0.6605168053866275-0.00015923170778380224j)

media_z0: None, line_z0: 50, Embed: False
 z0: 2257.60-2257.22j, Z0: 2257.60-2257.22j, gamma: 1.14+1.14j
 line S11 0j
 line S21 (0.988617463159112-0.011255830151580522j)
 line z0 (50+0j)

media_z0: None, line_z0: 50, Embed: True
 z0: 2257.60-2257.22j, Z0: 2257.60-2257.22j, gamma: 1.14+1.14j
 line S11 (-0.3394831573782804+9.291482605806173e-05j)
 line S21 (0.6605168053866376-0.00015923170763578625j)
 line z0 (2257.6025298862933-2257.2199592444235j)
 line renorm S11 (1.1379786002407855e-15-9.992007221626409e-15j)
 line renorm S21 (0.9886174631591118-0.011255830151576995j)

S-parameters calculated as series resistor
dc R 50.92958178940652
Resistor S11 (0.3374393620229403+0j)
Resistor S21 (0.6625606379770599+0j)
```

By default there is no way to get S-parameters that match the series resistor calculation. `media_z0: None, line_z0: 50, Embed: True` is close but S11 is negative which means negative series resistor and the port impedance is not 50 ohms. I assumed that setting `line_z0=50` would use 50 ohm characteristic impedance but I'm not sure that's the case since port impedance is left as characteristic impedance?

The first two results is with 50 ohm port impedance, but skrf seems to calculate the S-parameters as if the characteristic impedance of line was 50 ohms. When reference impedance is set to characteristic impedance the S-parameters are calculated correctly, however the port impedance is left as characteristic impedance. It can be renormalized to 50 ohms but either `traveling` or `pseudo` definition must be used as `power` gives incorrect results.

This is the renormalization result if `power` definition is used:
```
 line renorm S11 (0.9992631627440348+0.021882842141062064j)
 line renorm S21 (0.0004818427027186568-0.022134902799545834j)
```

`pseudo` renormalization gives results very close to `traveling`.

I think the first step to fixing this would be to decide what the various `z0` arguments should do. I think `traveling` renormalization should be used in line and it should give correct results when the starting impedance is the characteristic impedance, but I'm not completely sure about this. `power` however seems to not work correctly.

### PR DESCRIPTION
**Pull request by mhuser**
This PR is the second attempt to fix #651 

Ready for review.

Todo
===

- [x] Refactor all `media` for `z0_port`, `z0` scheme
- [x] Refactor waveguides `media` for `z0_override`, `z0` scheme
- [x] Make all tests pass and refine
- [x] Make all notebooks build
- [x] Update notebooks (check contents)
- [x] Update documentation
- [x] Extend documentation (e.g. by passing `z0` to `Media.line())`

### CODE REVIEW

File: doc/source/tutorials/Media.ipynb
Comment (jhillairet) [inline doc/source/tutorials/Media.ipynb:1]: > *port impedance* `z0` of the line network 

`z0_port` ?

**Review by jhillairet — APPROVED** [review]
Great work! I think the media definition it is much clearer with this PR. I didn't reviewed the notebook yet.

I wonder how much code will be broken after the change and if we should merge it to the branch v1.0 instead (question)

File: doc/source/tutorials/Media.ipynb
Comment (mhuser) [reply doc/source/tutorials/Media.ipynb:1]: > > _port impedance_ `z0` of the line network
> 
> `z0_port` ?

Well, unfortnately for the clarity, the *port impedance* of Network object is `z0`, while *characteristic impedance* is not stored.
I should add a remark in this sense.

**Comment by mhuser** [comment]
> I wonder how much code will be broken after the change and if we should merge it to the branch v1.0 instead (question)

Hum. The scikit-rf should of course works, but the issue is the existing user code.

However, if they call `Media.Z0` it will alias with `Media.z0` with a warning. If they use `Media.line(... embed=True, ...) it will emit a warning and still get a line, but maybe the impedance will surprise them depending the case. However they will still have lenght, attenuation and so on.

This give me this idea: the parameter `z0` could be kept in all Media constructors but emit a warning and wrap to `z0_port` (resp. `z0_override` for waveguides) if this last is None.

As such, only the port impedance would eventually change but old codes would continue to run with warnings.

We could also merge in 1.0, but the cost will be to merge with all upcoming commits until this version.

File: doc/source/tutorials/Media.ipynb
Comment (mhuser) [reply doc/source/tutorials/Media.ipynb:1]: Done.

**Comment by mhuser** [comment]
A deprecation warning is now emmited if `z0` is used in `Media` initialization. This helped to find some more `z0` remaining.
The `z0` fallback on `z0_port`, resp. `z0_override` if these are None.

**Comment by mhuser** [comment]
After having reflected a bit on it, I think there is some tradeoff here. If we keep the PR to merge within 1.0, it probably will not benefits from the extensive testing by many users on Master, with the risk to miss key issues that we later have to fix in 1.0.
On the other hand, breaking user code can act as a repellent. Here they will always get a media line with a warning in the cases were the behaviour is subject to have changed.

**Comment by jhillairet** [comment]
Updated documentation for reviewers:
https://scikit-rf--727.org.readthedocs.build/en/727/

- doc/source/tutorials/Introduction.ipynb
- doc/source/tutorials/Media.ipynb
- doc/source/examples/circuit/Wilkinson Power Splitter.ipynb
- doc/source/examples/interactive/Interact Transmission Lines.ipynb
- doc/source/examples/interactive/Interactive Coaxial Properties.ipynb
- doc/source/examples/interactive/Interactive Mismatched Line.ipynb
- doc/source/examples/metrology/Calibration With Three Receivers.ipynb
- doc/source/examples/metrology/LRRM.ipynb
- doc/source/examples/metrology/Multi-port Calibration.ipynb
- doc/source/examples/metrology/Multiline TRL.ipynb
- doc/source/examples/metrology/NanoVNA_V2_4port-splitter.ipynb
- doc/source/examples/metrology/SOLT Calibration Standards Creation.ipynb
- doc/source/examples/metrology/TwoPortOnePath, EnhancedResponse, and FakeFlip.ipynb
- doc/source/examples/networksets/Sorting Network Sets.ipynb
- doc/source/examples/networktheory/Balanced Network De-embedding.ipynb
- doc/source/examples/networktheory/CPW media example.ipynb
- ...ce/examples/networktheory/Correlating microstripline model to measurement.ipynb
- doc/source/examples/networktheory/DC Extrapolation for Time Domain .ipynb
- doc/source/examples/networktheory/DefinedAEpTandZ0 media example.ipynb
- doc/source/examples/networktheory/IEEEP370 Deembedding.ipynb
- doc/source/examples/networktheory/LNA Example.ipynb
- doc/source/examples/networktheory/Properties of Rectangular Waveguides.ipynb
- ...amples/networktheory/Time domain reflectometry, measurement vs simulation.ipynb
- ...rce/examples/networktheory/Transmission Line Properties and Manipulations.ipynb
- doc/source/examples/taper/tapered_transmission_lines.ipynb

File: doc/source/tutorials/Media.ipynb
[the remark this answers is not in the document]
Comment (jhillairet) [reply doc/source/tutorials/Media.ipynb:1]: Wouldn't be more logical to present media from the simplest (wrt to measurement point of view) to the hardest (those would requires adapters) ? I mean, medias like coex (without `z0_override` parameter) before waveguides and CPW?

File: skrf/media/media.py [hunk 1]
```diff
@@ -159,21 +162,21 @@ def npoints(self, val):
...
     @property
-    def z0(self) -> npy.ndarray:
+    def z0_port(self) -> npy.ndarray:
         """
         Characteristic Impedance.
 
         Returns
         -------
         z0 : :class:`numpy.ndarray`
         """
```
Comment (jhillairet) [inline skrf/media/media.py:172]: I think additional explanation should be given about what is returned here (None or something)

File: doc/source/tutorials/Media.ipynb
[the remark this answers is not in the document]
Comment (mhuser) [reply doc/source/tutorials/Media.ipynb:1]: Indeed, I think I was somehow wearing blinders due to refactorization focus.

File: skrf/media/coaxial.py [hunk 2]
```diff
@@ -35,10 +36,17 @@ class Coaxial(DistributedCircuit, Media):
...
+        impedance z0 of the media to `z0_port`.
+        Else if `z0_port` is None, the networks port impedances will be the raw
+        characteristic impedance z0 of the media.
+        (Default is None)
     z0 : number, array-like, or None
-        the port impedance for media. Only needed if its different
-        from the characteristic impedance of the transmission
-        line. if z0 is None then will default to Z0
+        deprecated parameter, alias to `z0_port` if `z0_port` is None.
+        Emmit a deprecation warning.
```
[the remark this answers is not in the document]
Comment (mhuser) [reply skrf/media/coaxial.py:49]: How about adding `z0_override` to all `Media`?  `z0` can then alias to `z0_override` in all cases, mimicking the old behaviour better.

File: skrf/media/circularWaveguide.py [hunk 3]
```diff
@@ -87,19 +100,33 @@ class CircularWaveguide(Media):
...
                  *args, **kwargs):
-
-        Media.__init__(self, frequency=frequency,z0=z0)
-
+        if z0 is not None:
+            # warns of deprecation
+            warnings.warn(
+                'Use of `z0` in Media init is deprecated.\n'
+                '`z0` alias with `z0_override if the last is not None`.\n'
+                '`z0` will be removed of Media init in version 1.0',
```
[the remark this answers is not in the document]
Comment (mhuser) [reply skrf/media/circularWaveguide.py:116]: DeprecationWarning: Use of `z0` in Media initialization is deprecated.
If `z0_override` is `None`, then `z0_override` will take `z0` value.
`z0` will be removed from Media initialization in version 1.0.

File: skrf/media/coaxial.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (jhillairet) [reply skrf/media/coaxial.py:49]: Yes, it would be logical to have the same behavior everywhere.

[the remark this answers is not in the document]
Comment (mhuser) [reply skrf/media/coaxial.py]: Done.

File: doc/source/tutorials/Media.ipynb
[the remark this answers is not in the document]
Comment (mhuser) [reply doc/source/tutorials/Media.ipynb:1]: Done, the tutorial now start by coaxial VNA-like measurement.

File: skrf/media/media.py (hunk 1 above)
Comment (mhuser) [reply skrf/media/media.py]: Now that `z0_override` is a property of the abstract class `Media`, the switch between `z0_override` and abstract method `z0_characteristic` can be done here. The media classes will need to define z0_characteristic, which sounds quite clear what it is 😉 
```Python
    @property
    @abstractmethod
    def z0_characteristic(self):
        """
        Characteristic Impedance.
        This abstract method has to be defined in the Media Class.
        
        Returns
        -------
        z0_characteristic: npy.ndarray
            Characteristic Impedance in units of ohms
        """
        return None
    
    @property
    def z0(self):
        """
        Return Characteristic Impedance `z0_characteristic`.
        If `z0_override` is not None, it is returned instead.
        
        Returns
        -------
        z0 : npy.ndarray
            z0_characteristic or z0_override in units of ohms
        """
        if self.z0_override is None:
            return self.z0_characteristic
        else:
            return self.z0_override
```

Comment (mhuser) [reply skrf/media/media.py]: Done.

**Comment by jhillairet** [comment]
@mhuser do you think it is ok for the next release now ? (or the one after the next one?)

**Comment by mhuser** [comment]
> @mhuser do you think it is ok for the next release now ? (or the one after the next one?)

I think yes. Let's merge branch master inside. 😅

**Review by jhillairet — APPROVED** [review]
Great (and huge) work @mhuser . Let's see if anyone complain after the next release ;)

**Comment by drsvenkuehn** [comment]
> This PR is the second attempt to fix #651
> 
> Ready for review.
> 
> # Todo
> * [x]  Refactor all `media` for `z0_port`, `z0` scheme
> * [x]  Refactor waveguides `media` for `z0_override`, `z0` scheme
> * [x]  Make all tests pass and refine
> * [x]  Make all notebooks build
> * [x]  Update notebooks (check contents)
> * [x]  Update documentation
> * [x]  Extend documentation (e.g. by passing `z0` to `Media.line())`

I believe this breaks the "media coaxial" example in the tutorial.

**Comment by jhillairet** [comment]
> I believe this breaks the "media coaxial" example in the tutorial.

@drsvenkuehn, which one ?
https://scikit-rf.readthedocs.io/en/latest/tutorials/Introduction.html#Coax ?
or 
https://scikit-rf.readthedocs.io/en/latest/tutorials/Media.html#RG58C/U-coaxial-cable ?

**Comment by drsvenkuehn** [comment]
@mhuser
 just realized that there is now 0.30.0. but the behaviour is the same for me.

**Comment by drsvenkuehn** [comment]
Hi @mhuser ...sorry, my bad. I didn't use the right virtual environment. I used and old version. I get the same results with 0.30.0

**Comment by jhillairet** [comment]
Ok, thanks for letting us know.

**Pull request merged**